### ISSUE TITLE
proposal: `<no_color_and_decoration>`

### ISSUE THREAD

**Issue by Fernandomr88** [dart-lang/linter#4805]
## Description

Whenever you're implementing a Container for example and you use both color and decoration at the same time, there's no error shown but an exception is thrown **only when the widget is built in runtime**, not even on compilation.
Code example:

```dart
Container(
  height: 150,
  width: 150,
  color: Colors.white, ///<- color
  decoration: BoxDecoration(borderRadius: BorderRadius.circular(100), color: Colors.white), //<- color again
  child: Image.asset(
    'assets/img/X.gif',
    height: 100,
  ),
));
```

![](https://user-images.githubusercontent.com/14304128/263733507-deefa638-e3f5-4a77-9428-f3fe2538a7be.png)

## Details

Cannot provide both a color and a decoration. To provide both, use `decoration: BoxDecoration(color: color)`

## Kind

Guard against runtime errors

## Bad Examples

```dart
Container(
  height: 150,
  width: 150,
  color: Colors.white, ///<- color
  decoration: BoxDecoration(borderRadius: BorderRadius.circular(100), color: Colors.white), //<- color again
  child: Image.asset(
    'assets/img/X.gif',
    height: 100,
  ),
));
```

## Good Examples 

```dart
Container(
  height: 150,
  width: 150,
  decoration: BoxDecoration(borderRadius: BorderRadius.circular(100), color: Colors.white),
  child: Image.asset(
    'assets/img/X.gif',
    height: 100,
  ),
));
```

## Discussion

This is coming from two previous created issues.

https://github.com/flutter/flutter/issues/133760
https://github.com/flutter/flutter/issues/133456

### Discussion checklist

I have come to this error multiple times, only in runtime - not even compile time, when I started to code using Flutter. This could have been prevented by this simple rule.

- [ ] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [ ] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [ ] If there's any prior art (e.g., in other linters), please add references here.
- [ ] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn't any corresponding advice, should there be?)
- [ ] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/main/doc/writing-lints.md
[Effective Dart]: https://dart.dev/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

**Comment by bwilkerson** [comment]
@goderbauer

**Comment by goderbauer** [comment]
It is correct that you can't have a color and decoration on a Container at the same time. Right now we can only detect this situation at runtime. A lint for this would be nice to have. However, this is not the only place in the API where we have mutually exclusive arguments. I would prefer it if we had a generic way of marking mutually exclusive arguments (and linting based on that) instead of implementing individual one-off lints for very specific cases.

**Comment by Fernandomr88** [comment]
> It is correct that you can't have a color and decoration on a Container at the same time. Right now we can only detect this situation at runtime. A lint for this would be nice to have. However, this is not the only place in the API where we have mutually exclusive arguments. I would prefer it if we had a generic way of marking mutually exclusive arguments (and linting based on that) instead of implementing individual one-off lints for very specific cases.

That would be much better indeed. Should I change the proposal?

**Comment by bwilkerson** [comment]
> I would prefer it if we had a generic way of marking mutually exclusive arguments ...

The first option that comes to mind (though maybe not the best) would be to define an annotation (like `@MutuallyExclusive('group name')`, only better) and allow the analyzer to warn if
- there's only one parameter in the group, or
- if there's an argument for two or more parameters in the same group.

A language feature might be nicer, but I don't know whether we could get that. (Also the annotation might be more flexible because any style of parameter could be in the same group.)

**Comment by Fernandomr88** [comment]
> A language feature might be nicer, but I don't know whether we could get that. (Also the annotation might be more flexible because any style of parameter could be in the same group.)

Agreed, but in the previous [issue](https://github.com/flutter/flutter/issues/133760#issuecomment-1703642144) regarding this, opened at flutter, I was instructed that it was not possible and it should be a linter rule. 

Making it a language feature would make more sense, but...

**Comment by bwilkerson** [comment]
If we define an annotation then we'll implement, warning rather than lints, and we can move this issue to the 'sdk' repository. But for now this is a reasonable place for the issue.

**Comment by pq** [comment]
💯 

I captured some related thoughts in https://github.com/dart-lang/sdk/issues/47712.

Input welcome!

**Comment by Fernandomr88** [comment]
> 💯
> 
> I captured some related thoughts in [dart-lang/sdk#47712](https://github.com/dart-lang/sdk/issues/47712).
> 
> Input welcome!

Wow, awesome to see that this is already a work in progress.
Thanks @pq

**Comment by pq** [comment]
Sure thing. To be clear though, it's just a proposal... Please add any context there. Any additional motivation is greatly valued!
